File: sdk/src/enforce.ts
/**
 * Wallet-side enforcement kit.
 *
 * Turns PaySafe from advisory into ENFORCED: a wrapped signer physically
 * refuses to sign an x402 payment authorization unless a fresh, Ed25519-
 * verified PaySafe allow-verdict exists for exactly that payment. Because the
 * verdict's payment commitment — sha256(network|pay_to|asset|amount|nonce) —
 * is recomputed from the typed data the wallet is being asked to sign, a
 * compromised agent cannot scan one payment and settle another: any change to
 * the recipient, amount, asset, chain, or nonce changes the commitment and
 * the signature is refused.
 *
 *   const paysafe  = new PaySafeClient({ agentId: "my-agent" });
 *   const enforcer = new PaySafeEnforcer({ trustedKeyHex: PINNED_KEY });
 *   const account  = enforcer.guardSigner(privateKeyToAccount(PRIVATE_KEY));
 *
 *   const scan = await paysafe.guardOutgoing(payment);  // throws on block
 *   enforcer.approve(scan, payment);                    // registers the verdict
 *   // ... hand `account` to your x402 client as usual. It will only ever
 *   // sign authorizations whose commitment carries a live allow-verdict.
 *
 * Design notes:
 *  - Zero dependencies, signer-agnostic: anything with a `signTypedData`
 *    method works (viem accounts/wallet clients, ethers v6 signers, custom).
 *    The wrapper is a Proxy — every other property/method passes through.
 *  - Recognized payment types: EIP-3009 TransferWithAuthorization /
 *    ReceiveWithAuthorization (the x402 "exact" scheme on EVM) and ERC-2612
 *    Permit. Other typed data passes through by default; set
 *    `strictTypes: true` to refuse everything unrecognized (deny-by-default).
 *  - Approvals are SINGLE-USE by default and expire with the attestation
 *    (plus an optional tighter `maxAgeMs`), so a verdict can't be hoarded.
 *  - Enforcement never phones home: approval happens locally against the
 *    pinned key. If PaySafe is unreachable, nothing new can be approved —
 *    fail-closed, which is the point.
 */
import {
  computePaymentCommitment,
  verifyAttestation,
  type PaymentDetails,
  type ScanResponse,
  type VerdictAttestation,
} from "./index.ts";

// ---------------------------------------------------------------------------
// Typed-data shapes (structural — no viem/ethers dependency)
// ---------------------------------------------------------------------------
export interface Eip712DomainLike {
  name?: string;
  version?: string;
  chainId?: number | bigint | string;
  verifyingContract?: string;
}

export interface TypedDataLike {
  domain?: Eip712DomainLike;
  types?: Record<string, unknown>;
  primaryType?: string;
  message?: Record<string, unknown>;
}

/** Anything with a signTypedData method (viem account, ethers signer, ...). */
export interface TypedDataSigner {
  signTypedData(...args: unknown[]): Promise<unknown>;
}

export interface EnforcerOptions {
  /** PINNED PaySafe verdict key (hex SPKI DER, from /.well-known/paysafe-verdict-key).
   * Required — enforcement without a pinned key would trust whatever key a
   * forged response embeds. */
  trustedKeyHex: string;
  /** Also accept "flag" verdicts (default false: allow-only). */
  allowFlagged?: boolean;
  /** Tighter freshness bound than the attestation's own expires_at, measured
   * from approve() time. Optional. */
  maxAgeMs?: number;
  /** Let one approval sign multiple times (default false: single-use). */
  reusable?: boolean;
  /** Refuse to sign ANY typed data that isn't a recognized payment
   * authorization (default false: unrecognized types pass through). */
  strictTypes?: boolean;
}

export class PaySafeEnforcementError extends Error {
  readonly commitment?: string;
  readonly primaryType?: string;
  constructor(message: string, info: { commitment?: string; primaryType?: string } = {}) {
    super(message);
    this.name = "PaySafeEnforcementError";
    this.commitment = info.commitment;
    this.primaryType = info.primaryType;
  }
}

interface Approval {
  attestation: VerdictAttestation;
  scanId: string;
  verdict: string;
  approvedAt: number;
  used: boolean;
}

/**
 * Map an EIP-712 payload to the payment fields the commitment binds.
 * Returns null for typed data that is not a recognized payment authorization.
 */
export function paymentFromTypedData(td: TypedDataLike): PaymentDetails | null {
  const pt = td.primaryType;
  if (!pt) return null;
  const m = td.message ?? {};
  const d = td.domain ?? {};
  const s = (v: unknown): string =>
    typeof v === "string" ? v : typeof v === "number" || typeof v === "bigint" ? String(v) : "";
  const network = d.chainId !== undefined && d.chainId !== null && s(d.chainId) !== "" ? `eip155:${s(d.chainId)}` : "";
  const asset = typeof d.verifyingContract === "string" ? d.verifyingContract : "";

  // EIP-3009 — how the x402 "exact" scheme authorizes USDC transfers on EVM.
  if (pt === "TransferWithAuthorization" || pt === "ReceiveWithAuthorization") {
    return { network, asset, pay_to: s(m.to), payer: s(m.from), amount: s(m.value), nonce: s(m.nonce) };
  }
  // ERC-2612 Permit — a spend approval is a payment authorization too.
  if (pt === "Permit") {
    return { network, asset, pay_to: s(m.spender), payer: s(m.owner), amount: s(m.value), nonce: s(m.nonce) };
  }
  return null;
}

export class PaySafeEnforcer {
  private readonly trustedKeyHex: string;
  private readonly allowFlagged: boolean;
  private readonly maxAgeMs: number | null;
  private readonly reusable: boolean;
  private readonly strictTypes: boolean;
  private readonly approvals = new Map<string, Approval>();

  constructor(opts: EnforcerOptions) {
    if (!opts.trustedKeyHex) {
      throw new PaySafeEnforcementError(
        "trustedKeyHex is required: pin the PaySafe verdict key (GET /.well-known/paysafe-verdict-key) — enforcement must never trust a key embedded in a response.",
      );
    }
    this.trustedKeyHex = opts.trustedKeyHex;
    this.allowFlagged = opts.allowFlagged ?? false;
    this.maxAgeMs = opts.maxAgeMs ?? null;
    this.reusable = opts.reusable ?? false;
    this.strictTypes = opts.strictTypes ?? false;
  }

  /**
   * Register a scan verdict as signing authority for its payment. Verifies the
   * attestation against the PINNED key (signature, field match, commitment,
   * expiry — throws AttestationError on any failure), then requires an allow
   * verdict (or flag with allowFlagged). Returns the payment commitment.
   */
  approve(scan: ScanResponse, payment: PaymentDetails): string {
    verifyAttestation(scan, payment, this.trustedKeyHex);
    if (scan.verdict !== "allow" && !(scan.verdict === "flag" && this.allowFlagged)) {
      throw new PaySafeEnforcementError(
        `refusing to approve a "${scan.verdict}" verdict for signing${scan.verdict === "flag" ? " (set allowFlagged to accept flags)" : ""}`,
      );
    }
    const commitment = computePaymentCommitment(payment);
    this.approvals.set(commitment, {
      attestation: scan.attestation as VerdictAttestation,
      scanId: scan.scan_id,
      verdict: scan.verdict,
      approvedAt: Date.now(),
      used: false,
    });
    return commitment;
  }

  /** Withdraw signing authority for a commitment. */
  revoke(commitment: string): void {
    this.approvals.delete(commitment);
  }

  /** Drop all approvals. */
  clear(): void {
    this.approvals.clear();
  }

  /**
   * The sign-time gate. Throws PaySafeEnforcementError unless a live,
   * unexpired (and unused, unless reusable) approval exists for this
   * commitment; consumes it on success.
   */
  assertApproved(commitment: string, primaryType?: string): void {
    const a = this.approvals.get(commitment);
    if (!a) {
      throw new PaySafeEnforcementError(
        `no PaySafe allow-verdict for this payment authorization (commitment ${commitment.slice(0, 16)}…). ` +
          `Scan the payment and call enforcer.approve(scan, payment) first. If the payment was scanned, its ` +
          `recipient/amount/asset/chain/nonce differs from what is now being signed — which is exactly what this gate exists to catch.`,
        { commitment, primaryType },
      );
    }
    if (a.used && !this.reusable) {
      throw new PaySafeEnforcementError(
        `this allow-verdict was already used to sign once (scan ${a.scanId}); approvals are single-use. Re-scan to sign again.`,
        { commitment, primaryType },
      );
    }
    const expiresAt = Date.parse(a.attestation.expires_at);
    const staleAt = this.maxAgeMs !== null ? a.approvedAt + this.maxAgeMs : Infinity;
    const deadline = Math.min(expiresAt, staleAt);
    if (Date.now() >= deadline) {
      this.approvals.delete(commitment);
      throw new PaySafeEnforcementError(
        `the allow-verdict for this payment is stale (scan ${a.scanId}, deadline ${new Date(deadline).toISOString()}). Re-scan to obtain a fresh verdict.`,
        { commitment, primaryType },
      );
    }
    a.used = true;
  }

  /**
   * Wrap a signer so signTypedData refuses x402/EIP-3009/Permit payment
   * authorizations that lack a live approval. Every other property and method
   * of the signer passes through untouched (Proxy), so the wrapped signer is
   * a drop-in replacement for viem accounts, ethers signers, etc.
   *
   * Scope: this guards the typed-data path x402 uses. If your signer exposes
   * other fund-moving paths (signTransaction), gate those at your policy
   * layer too.
   */
  guardSigner<T extends TypedDataSigner>(signer: T): T {
    // viem calls account.signTypedData(typedData); ethers v6 uses
    // signer.signTypedData(domain, types, message). Support both shapes.
    const enforcer = this;
    const wrapped = async function (this: unknown, ...args: unknown[]): Promise<unknown> {
      const td: TypedDataLike =
        args.length >= 3
          ? { domain: args[0] as Eip712DomainLike, types: args[1] as Record<string, unknown>, message: args[2] as Record<string, unknown>, primaryType: inferPrimaryType(args[1]) }
          : ((args[0] ?? {}) as TypedDataLike);
      const payment = paymentFromTypedData(td);
      if (payment === null) {
        if (enforcer.strictTypes) {
          throw new PaySafeEnforcementError(
            `strictTypes: refusing to sign unrecognized typed data (primaryType ${td.primaryType ?? "unknown"})`,
            { primaryType: td.primaryType },
          );
        }
        return signer.signTypedData(...args);
      }
      enforcer.assertApproved(computePaymentCommitment(payment), td.primaryType);
      return signer.signTypedData(...args);
    };
    return new Proxy(signer, {
      get(target, prop, receiver) {
        if (prop === "signTypedData") return wrapped;
        const value = Reflect.get(target, prop, receiver);
        return typeof value === "function" ? value.bind(target) : value;
      },
    });
  }
}

/** ethers v6 passes (domain, types, message) with no primaryType — infer it
 * as the type that no other type references (EIP-712 convention). */
function inferPrimaryType(types: unknown): string | undefined {
  if (typeof types !== "object" || types === null) return undefined;
  const names = Object.keys(types as Record<string, unknown>).filter((n) => n !== "EIP712Domain");
  const referenced = new Set<string>();
  for (const n of names) {
    const fields = (types as Record<string, unknown>)[n];
    if (!Array.isArray(fields)) continue;
    for (const f of fields) {
      const t = (f as { type?: string })?.type ?? "";
      referenced.add(t.replace(/\[\]$/, ""));
    }
  }
  return names.find((n) => !referenced.has(n)) ?? names[0];
}

File: sdk-python/src/paysafe_x402/enforce.py
"""
Wallet-side enforcement kit (Python parity with the TypeScript SDK's enforce.ts).

Turns PaySafe from advisory into ENFORCED: a wrapped signer refuses to sign an
x402 payment authorization unless a fresh, Ed25519-verified PaySafe
allow-verdict exists for exactly that payment. The verdict's payment
commitment — sha256(network|pay_to|asset|amount|nonce) — is recomputed from
the EIP-712 typed data the wallet is being asked to sign, so a compromised
agent cannot scan one payment and settle another: any change to recipient,
amount, asset, chain, or nonce changes the commitment and the signature is
refused.

    from paysafe_x402 import PaySafeClient, PaySafeEnforcer
    from eth_account import Account

    paysafe  = PaySafeClient(agent_id="my-agent")
    enforcer = PaySafeEnforcer(trusted_key_hex=paysafe.verdict_key())
    account  = enforcer.guard_signer(Account.from_key(PRIVATE_KEY))

    scan = paysafe.guard_outgoing(payment)      # raises on block
    enforcer.approve(scan, payment)             # registers the verdict locally
    # ... hand `account` to your x402 client as usual. It will only ever sign
    # authorizations whose commitment carries a live allow-verdict.

Design notes (identical guarantees to the TS kit):
  - Signer-agnostic: anything with a ``sign_typed_data`` method works —
    eth-account LocalAccount (positional, keyword, or ``full_message=`` call
    shapes are all recognized), web3.py wrappers, custom signers. Every other
    attribute/method of the signer passes through untouched.
  - Recognized payment types: EIP-3009 TransferWithAuthorization /
    ReceiveWithAuthorization (the x402 "exact" scheme on EVM) and ERC-2612
    Permit. Other typed data passes through by default; ``strict_types=True``
    refuses everything unrecognized (deny-by-default).
  - Approvals are SINGLE-USE by default and expire with the attestation (plus
    an optional tighter ``max_age_s``), so a verdict can't be hoarded.
  - Enforcement never phones home: approval happens locally against the
    pinned key. If PaySafe is unreachable, nothing new can be approved —
    fail-closed, which is the point.
"""
from __future__ import annotations

import time
from typing import Any, Dict, Optional

from . import (
    PaySafeError,
    _parse_iso_ms,
    compute_payment_commitment,
    verify_attestation,
)

__all__ = [
    "PaySafeEnforcer",
    "PaySafeEnforcementError",
    "payment_from_typed_data",
]

#: EIP-712 primary types treated as payment authorizations.
_PAYMENT_TYPES = ("TransferWithAuthorization", "ReceiveWithAuthorization", "Permit")


class PaySafeEnforcementError(PaySafeError):
    """Raised when a signature is refused (no/stale/used approval, strict mode)."""

    def __init__(self, message: str, commitment: Optional[str] = None, primary_type: Optional[str] = None):
        super().__init__(message)
        self.commitment = commitment
        self.primary_type = primary_type


def _s(v: Any) -> str:
    """Stringify typed-data scalar values the way the commitment expects."""
    if isinstance(v, str):
        return v
    if isinstance(v, bool):  # bool is an int subclass; exclude it explicitly
        return ""
    if isinstance(v, int):
        return str(v)
    if isinstance(v, (bytes, bytearray)):
        return "0x" + bytes(v).hex()
    return ""


def _infer_primary_type(types: Any) -> Optional[str]:
    """eth-account's (domain, types, message) shape has no primaryType — infer
    it as the type no other type references (EIP-712 convention)."""
    if not isinstance(types, dict):
        return None
    names = [n for n in types.keys() if n != "EIP712Domain"]
    referenced = set()
    for n in names:
        fields = types.get(n)
        if not isinstance(fields, list):
            continue
        for f in fields:
            t = (f or {}).get("type", "") if isinstance(f, dict) else ""
            referenced.add(t.replace("[]", ""))
    for n in names:
        if n not in referenced:
            return n
    return names[0] if names else None


def payment_from_typed_data(td: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """Map an EIP-712 payload to the payment fields the commitment binds.

    Returns None for typed data that is not a recognized payment authorization.
    """
    if not isinstance(td, dict):
        return None
    primary = td.get("primaryType") or _infer_primary_type(td.get("types"))
    if primary not in _PAYMENT_TYPES:
        return None
    message = td.get("message") or {}
    domain = td.get("domain") or {}
    chain_id = _s(domain.get("chainId"))
    network = f"eip155:{chain_id}" if chain_id else ""
    asset = domain.get("verifyingContract") if isinstance(domain.get("verifyingContract"), str) else ""

    if primary in ("TransferWithAuthorization", "ReceiveWithAuthorization"):
        return {
            "network": network,
            "asset": asset,
            "pay_to": _s(message.get("to")),
            "payer": _s(message.get("from")),
            "amount": _s(message.get("value")),
            "nonce": _s(message.get("nonce")),
        }
    # ERC-2612 Permit — a spend approval is a payment authorization too.
    return {
        "network": network,
        "asset": asset,
        "pay_to": _s(message.get("spender")),
        "payer": _s(message.get("owner")),
        "amount": _s(message.get("value")),
        "nonce": _s(message.get("nonce")),
    }


def _typed_data_from_call(args: tuple, kwargs: dict) -> Optional[Dict[str, Any]]:
    """Normalize the sign_typed_data call shapes into one EIP-712 dict.

    Recognized shapes:
      sign_typed_data(full_message={...})                       (eth-account kwarg)
      sign_typed_data({...})                                    (single full dict)
      sign_typed_data(domain_data, message_types, message_data) (eth-account positional)
      sign_typed_data(domain_data=..., message_types=..., message_data=...)
    """
    full = kwargs.get("full_message")
    if isinstance(full, dict):
        return full
    if len(args) == 1 and isinstance(args[0], dict) and (
        "types" in args[0] or "primaryType" in args[0] or "message" in args[0]
    ):
        return args[0]
    domain = kwargs.get("domain_data")
    types = kwargs.get("message_types")
    message = kwargs.get("message_data")
    if len(args) >= 3:
        domain, types, message = args[0], args[1], args[2]
    if domain is None and types is None and message is None:
        return None
    return {
        "domain": domain or {},
        "types": types or {},
        "message": message or {},
        "primaryType": _infer_primary_type(types),
    }


class _Approval:
    __slots__ = ("attestation", "scan_id", "verdict", "approved_at", "used")

    def __init__(self, attestation: Dict[str, Any], scan_id: str, verdict: str):
        self.attestation = attestation
        self.scan_id = scan_id
        self.verdict = verdict
        self.approved_at = time.time()
        self.used = False


class _GuardedSigner:
    """Proxy around a signer: gates sign_typed_data, passes everything else through."""

    def __init__(self, signer: Any, enforcer: "PaySafeEnforcer"):
        object.__setattr__(self, "_paysafe_signer", signer)
        object.__setattr__(self, "_paysafe_enforcer", enforcer)

    def __getattr__(self, name: str) -> Any:
        return getattr(object.__getattribute__(self, "_paysafe_signer"), name)

    def sign_typed_data(self, *args: Any, **kwargs: Any) -> Any:
        signer = object.__getattribute__(self, "_paysafe_signer")
        enforcer: PaySafeEnforcer = object.__getattribute__(self, "_paysafe_enforcer")
        td = _typed_data_from_call(args, kwargs)
        payment = payment_from_typed_data(td) if td is not None else None
        if payment is None:
            if enforcer.strict_types:
                primary = (td or {}).get("primaryType") if isinstance(td, dict) else None
                raise PaySafeEnforcementError(
                    f"strict_types: refusing to sign unrecognized typed data (primaryType {primary or 'unknown'})",
                    primary_type=primary,
                )
            return signer.sign_typed_data(*args, **kwargs)
        primary = (td or {}).get("primaryType") if isinstance(td, dict) else None
        enforcer.assert_approved(compute_payment_commitment(payment), primary)
        return signer.sign_typed_data(*args, **kwargs)


class PaySafeEnforcer:
    """Local, fail-closed signing authority backed by PaySafe allow-verdicts."""

    def __init__(
        self,
        trusted_key_hex: str,
        allow_flagged: bool = False,
        accept_overrides: bool = False,
        max_age_s: Optional[float] = None,
        reusable: bool = False,
        strict_types: bool = False,
    ):
        if not trusted_key_hex:
            raise PaySafeEnforcementError(
                "trusted_key_hex is required: pin the PaySafe verdict key "
                "(GET /.well-known/paysafe-verdict-key) — enforcement must never "
                "trust a key embedded in a response."
            )
        self.trusted_key_hex = trusted_key_hex
        self.allow_flagged = allow_flagged
        # OPT-IN: an agent that holds its own API key can point the approval
        # webhook at itself and "approve" its own flags; human-approved
        # "override:allow" verdicts are only trustworthy when the webhook
        # receiver is out of the agent's reach.
        self.accept_overrides = accept_overrides
        self.max_age_s = max_age_s
        self.reusable = reusable
        self.strict_types = strict_types
        self._approvals: Dict[str, _Approval] = {}

    def approve(self, scan: Dict[str, Any], payment: Dict[str, Any]) -> str:
        """Register a scan verdict as signing authority for its payment.

        Verifies the attestation against the PINNED key (signature, field
        match, commitment, expiry — raises AttestationError on any failure),
        then requires an allow verdict (or flag with allow_flagged). Returns
        the payment commitment.
        """
        verify_attestation(scan, payment, self.trusted_key_hex)
        verdict = scan.get("verdict")
        acceptable = (
            verdict == "allow"
            or (verdict == "flag" and self.allow_flagged)
            or (verdict == "override:allow" and self.accept_overrides)
        )
        if not acceptable:
            if verdict == "flag":
                hint = " (set allow_flagged=True to accept flags)"
            elif verdict == "override:allow":
                hint = " (human-approved overrides require the accept_overrides opt-in - see its security note)"
            else:
                hint = ""
            raise PaySafeEnforcementError(f'refusing to approve a "{verdict}" verdict for signing{hint}')
        commitment = compute_payment_commitment(payment)
        self._approvals[commitment] = _Approval(
            attestation=scan["attestation"], scan_id=scan.get("scan_id", ""), verdict=str(verdict)
        )
        return commitment

    def revoke(self, commitment: str) -> None:
        """Withdraw signing authority for a commitment."""
        self._approvals.pop(commitment, None)

    def clear(self) -> None:
        """Drop all approvals."""
        self._approvals.clear()

    def assert_approved(self, commitment: str, primary_type: Optional[str] = None) -> None:
        """The sign-time gate. Raises PaySafeEnforcementError unless a live,
        unexpired (and unused, unless reusable) approval exists for this
        commitment; consumes it on success."""
        approval = self._approvals.get(commitment)
        if approval is None:
            raise PaySafeEnforcementError(
                f"no PaySafe allow-verdict for this payment authorization (commitment {commitment[:16]}…). "
                "Scan the payment and call enforcer.approve(scan, payment) first. If the payment was scanned, "
                "its recipient/amount/asset/chain/nonce differs from what is now being signed — which is "
                "exactly what this gate exists to catch.",
                commitment=commitment,
                primary_type=primary_type,
            )
        if approval.used and not self.reusable:
            raise PaySafeEnforcementError(
                f"this allow-verdict was already used to sign once (scan {approval.scan_id}); "
                "approvals are single-use. Re-scan to sign again.",
                commitment=commitment,
                primary_type=primary_type,
            )
        expires_ms = _parse_iso_ms(approval.attestation["expires_at"])
        stale_ms = (approval.approved_at + self.max_age_s) * 1000 if self.max_age_s is not None else float("inf")
        deadline_ms = min(expires_ms, stale_ms)
        if time.time() * 1000 >= deadline_ms:
            del self._approvals[commitment]
            raise PaySafeEnforcementError(
                f"the allow-verdict for this payment is stale (scan {approval.scan_id}). "
                "Re-scan to obtain a fresh verdict.",
                commitment=commitment,
                primary_type=primary_type,
            )
        approval.used = True

    def guard_signer(self, signer: Any) -> Any:
        """Wrap a signer so sign_typed_data refuses x402/EIP-3009/Permit payment
        authorizations that lack a live approval. Every other attribute and
        method of the signer passes through untouched, so the wrapped signer
        is a drop-in replacement.

        Scope: this guards the typed-data path x402 uses. If your signer
        exposes other fund-moving paths (sign_transaction), gate those at your
        policy layer too.
        """
        return _GuardedSigner(signer, self)
